Add test controller implementation

adisonlampert opened the pull request:
### Motivation

Closes #498 
Closes #382

<img width="599" alt="Screenshot 2023-04-14 at 2 07 41 PM" src="https://user-images.githubusercontent.com/38566184/232123449-80de2463-67ac-4036-8aae-be6f4f38d4a7.png">

A continuation of #506 to use the VS Code Test Explorer 

Update: After talking with Vini, we decided not to implement opening the terminal on run and outputting the test results.

### Implementation

- Created a new class `TestController`
- Added middleware to create a VS Code [`TestItem`](https://code.visualstudio.com/api/references/vscode-api#TestItem) for each test in an open file.
- The test items are added to a [`TestController`](https://code.visualstudio.com/api/references/vscode-api#TestController) which is the entry point to discover and execute tests.
- When you either run the test from the gutter of the file editor tab or from the Test Explorer tab, executes the method `runHandler` using the [`TestRunProfile`](https://code.visualstudio.com/api/references/vscode-api#TestRunProfile).
- Updated the method registered as `rubyLsp.runTest` to `runOnClick` which finds the `TestItem` associated with the name of the test, opens the Test Explorer to display that item, and runs the test using the `runHandler` method.

### Manual Tests

1. Debug this branch on `ruby-lsp`
2. Open up a test file (ex. `test/document_test.rb`)
3. Observe there is a "Run" button next to the class definition and test method definitions
4. Open the Test Explorer tab, observe the class and test methods are runnable there
5. Click the "Run" button next to a test method; obverse that it passes
7. Click the "Run" button in the Test Explorer for the class; observe that all the tests are run and pass
8. Change one of the tests so that it will fail
9. Click the "Run" text above the method definition; obverse that it fails and that a failure message appears

vinistock reviewed (approved):
This looks amazing.

Is there a way for us to show output for tests that pass? Imagine you're trying to debug something and go
```ruby
def test_something
  # Can we see the output somewhere in the test explorer?
  puts variable_im_trying_to_understand
end
```

vinistock commented:
In fact, I think there's a panel for test results. We might be able to reveal that panel automatically to show output.

adisonlampert commented:
I updated `runHandler` so that it only runs test methods which solved the issue I described above. 

In the VS Code Testing API Guide, [the recommendation](https://code.visualstudio.com/api/extension-guides/testing#:~:text=The%20TestItem%20interface%20is%20simple%20and%20doesn%27t%20have%20room%20for%20custom%20data.%20If%20you%20need%20to%20associate%20extra%20information%20with%20a%20TestItem%2C%20you%20can%20use%20a%20WeakMap%3A) is to use `WeakMap` to add additional information to `TestItem` objects. Previously, we were using the `TestItem` description to store the test command (left version). I implemented `WeakMap` (right version). I feel like the right version is better because you can see the time it took for the test to run after it has finished and looks cleaner + I feel that it is more in line with the intended use of `TestItem`.

<p float="left">
<img width="362" alt="Screenshot 2023-04-18 at 12 07 23 PM" src="https://user-images.githubusercontent.com/38566184/232841650-db5edf67-3166-4c25-833d-5b14e57bbcb8.png">
<img width="360" alt="Screenshot 2023-04-18 at 12 05 02 PM" src="https://user-images.githubusercontent.com/38566184/232841620-1acea7b7-761e-4b43-951e-1c1539f65f7b.png">
</p>

vinistock commented:
It seems that every time a test runs, a new test output tab is created. It is possible to reuse the same one?

adisonlampert commented on src/testController.ts:
```diff
@@ -0,0 +1,207 @@
+    request: vscode.TestRunRequest,
+    token: vscode.CancellationToken
+  ) {
+    const run = this.testController.createTestRun(request, undefined, true);
```
@vinistock 

The same output terminal was being used for tests run through the Test Explorer UI, but a new terminal was created each time a test was run through code lens (i.e., through `rubyLsp.runTest`). 

Since I am using `run.appendOutput` to log the test results, I knew that `run` must contain some reference to a terminal and that terminal was the same for anything run through the Test Explorer UI but not for code lens runs. 

By debugging, I observed that every Text Explorer `TestRun` had `persisted?` set to `true` even though we were setting it to false on line 100. However, every code lens `TestRun` had `persisted?` set to `false`. This was the only difference I could find between the two runs so I set `persisted?` to true here and that fixed the issue of opening multiple terminals.

adisonlampert commented on src/testController.ts:
```diff
@@ -0,0 +1,207 @@
+      true
+    );
+
+    vscode.commands.executeCommand("testing.clearTestResults");
```
I added in this line to clear the test results each time the LSP is started/restarted.

BTW, `testing.clearTestResults` and `testing.showMostRecentOutput` are not in the VS Code API documentation (or I couldn't find them in there). I got these commands from the source code [here](https://github.com/microsoft/vscode/blob/8f74fbfd1f2d8f6268a42df131726b218aafe511/src/vs/workbench/contrib/testing/common/constants.ts#L53). There are other test-related commands that might be useful later so I wanted to document it.

vinistock commented on src/testController.ts:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  createTestItems(response: CodeLens[]) {
+    this.testController.items.forEach((test) => {
+      this.testController.items.delete(test.id);
+    });
```
Should we also clear the `this.testCommands` weakmap here?

paracycle commented on src/client.ts:
```diff
@@ -117,6 +125,18 @@ export default class Client implements ClientInterface {
       },
       middleware: {
+        provideCodeLenses: async (document, token, next) => {
+          if (this.client) {
+            const response = await next(document, token);
+
+            if (response) {
+              this.testController.createTestItems(response as CodeLens[]);
+            }
+
+            return response;
+          }
+          return null;
```
Nitpick, can we invert the check to reduce nesting here?
```suggestion
          if (!this.client) {
            return null;
          }
          
          const response = await next(document, token);

          if (response) {
            this.testController.createTestItems(response as CodeLens[]);
          }

          return response;
```

paracycle commented on src/testController.ts:
```diff
@@ -0,0 +1,197 @@
+
+    this.testCommands = new WeakMap<vscode.TestItem, string>();
+
+    this.testRunProfile = this.testController.createRunProfile(
```
Can we have more than 1 profile defined? Could we also have a run profile that enables debugging? (Probably not as part of this PR, though)

vinistock replied:
Yeah, we intend to create a debugging profile to launch debugging sessions from tests.

paracycle commented on src/client.ts:
```diff
@@ -51,6 +54,11 @@ export default class Client implements ClientInterface {
     this.statusItems = new StatusItems(this);
     this.registerCommands();
     this.registerAutoRestarts();
+    this.testController = new TestController(
+      context,
+      this.workingFolder,
+      this.ruby
+    );
```
I don't think the `TestController` creation should be the responsibility of the `Client` class, which is supposed to be just the LSP client. I think we can move the creation of the test controller to `extension.ts` and inject it to the client as needed.

paracycle commented on src/client.ts:
```diff
@@ -117,6 +125,18 @@ export default class Client implements ClientInterface {
+            const response = await next(document, token);
+
+            if (response) {
+              this.testController.createTestItems(response as CodeLens[]);
```
Is this assuming that all code lenses will be testing related? What happens if the LSP server decides to add other kinds of code lenses? Will this code break?

vinistock reviewed (approved):
🚀

danielpgross commented:
This is really exciting, great work! 🎉
